src/tables.jl

```diff
@@ -1,42 +1,127 @@
+
+# Rows
+
 Tables.rowaccess(::Type{<:Result}) = true
 Tables.rows(jl_result::Result) = jl_result
 
 Base.eltype(jl_result::Result) = Row
 Base.length(jl_result::Result) = num_rows(jl_result)
 
+function Base.iterate(jl_result::Result, (len, row)=(length(jl_result), 1))
+    row > len && return nothing
+    return Row(jl_result, row), (len, row + 1)
+end
+
 function Tables.schema(jl_result::Result)
     types = map(jl_result.not_null, column_types(jl_result)) do not_null, col_type
         not_null ? col_type : Union{col_type, Missing}
     end
     return Tables.Schema(map(Symbol, column_names(jl_result)), types)
 end
 
-function Base.iterate(jl_result::Result, (len, row)=(length(jl_result), 1))
-    row > len && return nothing
-    return Row(jl_result, row), (len, row + 1)
-end
-
 struct Row
     result::Result
     row::Int
 end
 
-Base.propertynames(r::Row) = column_names(getfield(r, :result))
+result(pqrow::Row) = getfield(pqrow, :result)
+row_number(pqrow::Row) = getfield(pqrow, :row)
+
+Base.propertynames(pqrow::Row) = map(Symbol, column_names(result(pqrow)))
 
 function Base.getproperty(pqrow::Row, name::Symbol)
-    jl_result = getfield(pqrow, :result)
-    row = getfield(pqrow, :row)
+    jl_result = result(pqrow)
+    row = row_number(pqrow)
     col = column_number(jl_result, name)
-    if libpq_c.PQgetisnull(jl_result.result, row - 1, col - 1) == 1
+    return jl_result[row, col]
+end
+
+function Base.getindex(pqrow::Row, col::Integer)
+    row = row_number(pqrow)
+    return result(pqrow)[row, col]
+end
+
+Base.length(pqrow::Row) = num_columns(result(pqrow))
+
+function Base.iterate(pqrow::Row, (len, col)=(length(pqrow), 1))
+    col > len && return nothing
+    return (result(pqrow)[row_number(pqrow), col], (len, col + 1))
+end
+
+# Columns
+
+struct Column{T} <: AbstractVector{T}
+    result::Result
+    col::Int
+    col_name::Symbol
+    oid::Oid
+    not_null::Bool
+    typ::Type
+    func::Base.Callable
+end
+
+struct Columns <: AbstractVector{Column}
+    result::Result
+end
+
+result(cs::Columns) = getfield(cs, :result)
+
+Base.propertynames(cs::Columns) = map(Symbol, column_names(result(cs)))
+
+Base.getproperty(cs::Columns, name::Symbol) = Column(result(cs), name)
+Base.getindex(cs::Columns, col::Integer) = Column(result(cs), col)
+Base.IndexStyle(::Type{Columns}) = IndexLinear()
+Base.length(cs::Columns) = num_columns(result(cs))
+Base.size(cs::Columns) = (length(cs),)
+
+Tables.columnaccess(::Type{<:Result}) = true
+Tables.columns(jl_result::Result) = Columns(jl_result)
+
+function Tables.schema(cs::Columns)
+    jl_result = result(cs)
+    types = map(jl_result.not_null, column_types(jl_result)) do not_null, col_type
+        not_null ? col_type : Union{col_type, Missing}
+    end
+    return Tables.Schema(map(Symbol, column_names(jl_result)), types)
+end
+
+function Column(jl_result::Result, col::Integer, name=Symbol(column_name(jl_result, col)))
+    if !checkindex(Bool, 1:num_columns(jl_result), col)
```

Comment:
Would it be worth wrapping this in `@boundscheck` so that it can be elided if the `Column` constructor is inlined and called with `@inbounds`? (Not that that's likely to happen...)

Reply:
Yeah.

I wish I knew what `@propagate_inbounds` is for though.

Reply:
I too wish that.